networking_dvs/plugins/ml2/drivers/mech_dvs/agent/dvs_firewall.py

```diff
@@ -17,28 +20,46 @@
 class DvsSecurityGroupsDriver(firewall.FirewallDriver):
     def __init__(self, integration_bridge=None):
         self.v_center = integration_bridge if isinstance(integration_bridge, VCenter) else VCenter(self.conf.ML2_VMWARE)
-        self._defer_apply = False
         self._ports_by_device_id = {}  # Device-id seems to be the same as port id
+        self._sg_aggregates_per_dvs_uuid = defaultdict(lambda : defaultdict(dict))
 
     def prepare_port_filter(self, ports):
-        self._process_port_filter(ports)
+        # LOG.debug("prepare_port_filter called with %s", pprint.pformat(ports))
+        merged_ports = self._merge_port_info_from_vcenter(ports)
+        self._update_ports_by_device_id(merged_ports)
+        self._process_ports(merged_ports)
+        self._apply_changed_sg_attr()
+        self._reassign_ports(merged_ports)
 
     def apply_port_filter(self, ports):
-        self._process_port_filter(ports)
+        # This driver does all of its processing during the prepare_port_filter call
+        pass
 
     def update_port_filter(self, ports):
-        self._process_port_filter(ports)
+        # LOG.debug("update_port_filter called with %s", pprint.pformat(ports))
+        ports_to_remove = [self._ports_by_device_id[port['device']] for port in ports]
+        self._process_ports(ports_to_remove, decrement=True)
+
+        merged_ports = self._merge_port_info_from_vcenter(ports)
+        self._update_ports_by_device_id(merged_ports)
+
+        self._process_ports(merged_ports)
+        self._apply_changed_sg_attr()
+        self._reassign_ports(merged_ports)
 
     def remove_port_filter(self, port_ids):
-        self._remove_sg_from_dvs_port(port_ids)
+        # LOG.debug("remote_port_filter called for %s", pprint.pformat(port_ids))
+        ports_to_remove = [self._ports_by_device_id[port_id] for port_id in port_ids]
```

Comment:
[observed during tests on staging] we can get rpc messages from the queue that cause a failure here because _ports_by_device_id is empty on agent boot.